arbos/constraints/multi_dimensional.go

```diff
@@ -0,0 +1,213 @@
+}
+
+// OpenMultiGasConstraint opens or initializes a constraint in the given storage subspace.
+func OpenMultiGasConstraint(sto *storage.Storage) *MultiGasConstraint {
+	c := &MultiGasConstraint{
+		target:           sto.OpenStorageBackedUint64(targetOffset),
+		adjustmentWindow: sto.OpenStorageBackedUint64(adjustmentWindowOffset),
+		backlog:          sto.OpenStorageBackedUint64(backlogOffset),
+		sumWeights:       sto.OpenStorageBackedUint64(sumWeightsOffset),
+	}
+	for i := range int(multigas.NumResourceKind) {
+		// #nosec G115 safe: NumResourceKind < 2^32
+		offset := weightedResourcesBaseOffset + uint64(i)
+		c.weightedResources[i] = sto.OpenStorageBackedUint64(offset)
+	}
+	return c
+}
+
+// Clear resets the constraint and all weighted resources.
+func (c *MultiGasConstraint) Clear() error {
+	if err := c.target.Clear(); err != nil {
+		return err
+	}
+	if err := c.adjustmentWindow.Clear(); err != nil {
+		return err
+	}
+	if err := c.backlog.Clear(); err != nil {
+		return err
+	}
+	if err := c.sumWeights.Clear(); err != nil {
+		return err
+	}
+	for i := range int(multigas.NumResourceKind) {
+		if err := c.weightedResources[i].Clear(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// SetResourceWeights assigns per-resource weight multipliers for this constraint.
+func (c *MultiGasConstraint) SetResourceWeights(weights map[uint8]uint64) error {
+	var total uint64
+	for kind, weight := range weights {
+		if _, err := multigas.CheckResourceKind(kind); err != nil {
+			return err
+		}
+		total = arbmath.SaturatingUAdd(total, weight)
+	}
+	for i := range int(multigas.NumResourceKind) {
+		// #nosec G115 safe: NumResourceKind < 2^32
+		weight, ok := weights[uint8(i)]
+		if ok {
+			if err := c.weightedResources[i].Set(weight); err != nil {
+				return err
+			}
+		} else {
+			// zeroise weight for unspecified resources
+			if err := c.weightedResources[i].Set(0); err != nil {
+				return err
+			}
+		}
+	}
+	return c.sumWeights.Set(total)
+}
+
+// ConstraintContribution returns the normalized backlog contribution for a specific resource kind.
+func (c *MultiGasConstraint) ConstraintContribution(kind uint8) (uint64, error) {
+	if _, err := multigas.CheckResourceKind(kind); err != nil {
+		return 0, err
+	}
+
+	weight, err := c.weightedResources[int(kind)].Get()
+	if err != nil {
+		return 0, err
+	}
+	if weight == 0 {
+		return 0, nil
+	}
+
+	backlog, err := c.backlog.Get()
+	if err != nil || backlog == 0 {
+		return 0, err
+	}
+
+	target, err := c.target.Get()
+	if err != nil || target == 0 {
+		return 0, err
+	}
+
+	adjustmentWindow, err := c.adjustmentWindow.Get()
+	if err != nil || adjustmentWindow == 0 {
+		return 0, err
+	}
+
+	sumWeights, err := c.sumWeights.Get()
+	if err != nil || sumWeights == 0 {
+		return 0, err
+	}
+
+	contrib := arbmath.SaturatingUMul(backlog, weight) /
+		arbmath.SaturatingUMul(adjustmentWindow,
+			arbmath.SaturatingUMul(target, sumWeights),
+		)
```

Comment:
```suggestion
	dividend := arbmath.NaturalToBips(
		arbmath.SaturatingCast[int64](arbmath.SaturatingUMul(backlog, weight))
	divisor := arbmath.SaturatingCastToBips(
		arbmath.SaturatingUMul(adjustmentWindow,
			arbmath.SaturatingUMul(target, sumWeights)))
	exponent := dividend / divisor
```
We should do this division in Bips, like the existing model. See https://github.com/OffchainLabs/nitro/blob/master/arbos/l2pricing/model.go#L129